<a href="https://colab.research.google.com/github/Hnin225/HH2/blob/main/huggingface_transformer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from transformers import pipeline

camembert_fill_mask = pipeline("fill-mask", model="camembert-base")
results = camembert_fill_mask("Le camembert est <mask> :)")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/508 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/445M [00:00<?, ?B/s]

Some weights of the model checkpoint at camembert-base were not used when initializing CamembertForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing CamembertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing CamembertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/811k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.40M [00:00<?, ?B/s]

In [ ]:
results

[{'score': 0.5239005088806152,
  'token': 7200,
  'token_str': 'délicieux',
  'sequence': 'Le camembert est délicieux :)'},
 {'score': 0.09637847542762756,
  'token': 2183,
  'token_str': 'excellent',
  'sequence': 'Le camembert est excellent :)'},
 {'score': 0.036324553191661835,
  'token': 26202,
  'token_str': 'succulent',
  'sequence': 'Le camembert est succulent :)'},
 {'score': 0.029080098494887352,
  'token': 528,
  'token_str': 'meilleur',
  'sequence': 'Le camembert est meilleur :)'},
 {'score': 0.027218418195843697,
  'token': 1654,
  'token_str': 'parfait',
  'sequence': 'Le camembert est parfait :)'}]

In [1]:
from transformers import pipeline

bert_fill_mask = pipeline("fill-mask", model="bert-base-uncased")
results = bert_fill_mask("I wake up in the morning and eat [MASK].") # Changed <mask> to [MASK]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

BertForMaskedLM has generative capabilities, as `prepare_inputs_for_generation` is explicitly overwritten. However, it doesn't directly inherit from `GenerationMixin`. From 👉v4.50👈 onwards, `PreTrainedModel` will NOT inherit from `GenerationMixin`, and this model will lose the ability to call `generate` and other related functions.
  - If you're using `trust_remote_code=True`, you can get rid of this warning by loading the model with an auto class. See https://huggingface.co/docs/transformers/en/model_doc/auto#auto-classes
  - If you are the owner of the model architecture code, please modify your model class such that it inherits from `GenerationMixin` (after `PreTrainedModel`, otherwise you'll get an exception).
  - If you are not the owner of the model architecture class, please contact the model code owner to update it.
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'c

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [ ]:
results

[{'score': 0.43051397800445557,
  'token': 6350,
  'token_str': 'breakfast',
  'sequence': 'i wake up in the morning and eat breakfast.'},
 {'score': 0.10028998553752899,
  'token': 4596,
  'token_str': 'dinner',
  'sequence': 'i wake up in the morning and eat dinner.'},
 {'score': 0.07404348254203796,
  'token': 2498,
  'token_str': 'nothing',
  'sequence': 'i wake up in the morning and eat nothing.'},
 {'score': 0.03402998298406601,
  'token': 2673,
  'token_str': 'everything',
  'sequence': 'i wake up in the morning and eat everything.'},
 {'score': 0.03320090472698212,
  'token': 6265,
  'token_str': 'lunch',
  'sequence': 'i wake up in the morning and eat lunch.'}]

In [2]:
# prompt: 미얀마어로도 해줘

from transformers import pipeline

# Korean example
korean_fill_mask = pipeline("fill-mask", model="klue/bert-base")
results = korean_fill_mask("아침에 일어나서 [MASK]를 먹습니다.")
print("Korean results:")
print(results)

# Myanmar example (requires a suitable model)
# Currently, there's no readily available, high-quality pre-trained fill-mask model specifically for Myanmar.
# You'd need to find a Myanmar language model and fine-tune it for fill-mask tasks.
# Here's how you'd use it if you had such a model:

# Replace "myanmar-fill-mask-model" with the actual model name
# myanmar_fill_mask = pipeline("fill-mask", model="myanmar-fill-mask-model")
# results = myanmar_fill_mask("မနက်ဖြန် [MASK] သွားမယ်။") # Example Myanmar sentence
# print("\nMyanmar results:")
# print(results)


# For now, let's use a multilingual model and see if it works.
xlm_fill_mask = pipeline("fill-mask", model="xlm-roberta-base")
results = xlm_fill_mask("မနက်ဖြန် [MASK] သွားမယ်။", top_k=5) #Example Myanmar sentence
print("\nMultilingual Model Results (may not be accurate):")
results

config.json:   0%|          | 0.00/425 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/445M [00:00<?, ?B/s]

Some weights of the model checkpoint at klue/bert-base were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/289 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/248k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/495k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

Korean results:
[{'score': 0.12011479586362839, 'token': 13309, 'token_str': '샌드위치', 'sequence': '아침에 일어나서 샌드위치 를 먹습니다.'}, {'score': 0.0541873499751091, 'token': 14995, 'token_str': '햄버거', 'sequence': '아침에 일어나서 햄버거 를 먹습니다.'}, {'score': 0.05142130330204964, 'token': 18766, 'token_str': '브런치', 'sequence': '아침에 일어나서 브런치 를 먹습니다.'}, {'score': 0.04145331680774689, 'token': 9106, 'token_str': '샐러드', 'sequence': '아침에 일어나서 샐러드 를 먹습니다.'}, {'score': 0.0370374470949173, 'token': 7282, 'token_str': '우유', 'sequence': '아침에 일어나서 우유 를 먹습니다.'}]


config.json:   0%|          | 0.00/615 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.12G [00:00<?, ?B/s]

Some weights of the model checkpoint at xlm-roberta-base were not used when initializing XLMRobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing XLMRobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing XLMRobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.10M [00:00<?, ?B/s]

PipelineException: No mask_token (<mask>) found on the input

In [3]:
# prompt: 미얀마어로도 해줘

from transformers import pipeline

# ... (previous code)

# For now, let's use a multilingual model and see if it works.
xlm_fill_mask = pipeline("fill-mask", model="xlm-roberta-base")
# Change [MASK] to <mask>
results = xlm_fill_mask("မနက်ဖြန် <mask> သွားမယ်။", top_k=5) # Example Myanmar sentence
print("\nMultilingual Model Results (may not be accurate):")
results

Some weights of the model checkpoint at xlm-roberta-base were not used when initializing XLMRobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing XLMRobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing XLMRobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).



Multilingual Model Results (may not be accurate):


[{'score': 0.10281659662723541,
  'token': 95524,
  'token_str': 'ပြန်',
  'sequence': 'မနက်ဖြန် ပြန် သွားမယ်။'},
 {'score': 0.08495812863111496,
  'token': 9547,
  'token_str': 'ပြီး',
  'sequence': 'မနက်ဖြန်ပြီး သွားမယ်။'},
 {'score': 0.05394543334841728,
  'token': 13172,
  'token_str': 'ပဲ',
  'sequence': 'မနက်ဖြန်ပဲ သွားမယ်။'},
 {'score': 0.044581640511751175,
  'token': 3539,
  'token_str': 'ပါ',
  'sequence': 'မနက်ဖြန်ပါ သွားမယ်။'},
 {'score': 0.04113183170557022,
  'token': 7457,
  'token_str': 'မှာ',
  'sequence': 'မနက်ဖြန်မှာ သွားမယ်။'}]

In [ ]:
import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

sequence = "I've been waiting for a HuggingFace course my whole life."

tokens = tokenizer.tokenize(sequence)
ids = tokenizer.convert_tokens_to_ids(tokens)

input_ids = torch.tensor([ids])  # 배치 차원을 추가
output = model(input_ids)


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

In [ ]:
tokens

['i',
 "'",
 've',
 'been',
 'waiting',
 'for',
 'a',
 'hugging',
 '##face',
 'course',
 'my',
 'whole',
 'life',
 '.']

In [ ]:
ids

[1045,
 1005,
 2310,
 2042,
 3403,
 2005,
 1037,
 17662,
 12172,
 2607,
 2026,
 2878,
 2166,
 1012]

In [ ]:
input_ids

tensor([[ 1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,  2607,
          2026,  2878,  2166,  1012]])

In [ ]:
sequences = ["Hello, world!", "Hugging Face is amazing!"]
batch = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")
output = model(**batch)

In [ ]:
batch

{'input_ids': tensor([[  101,  7592,  1010,  2088,   999,   102,     0],
        [  101, 17662,  2227,  2003,  6429,   999,   102]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 0],
        [1, 1, 1, 1, 1, 1, 1]])}

In [ ]:
batch['input_ids']

tensor([[  101,  7592,  1010,  2088,   999,   102,     0],
        [  101, 17662,  2227,  2003,  6429,   999,   102]])

In [ ]:
batch['attention_mask']

tensor([[1, 1, 1, 1, 1, 1, 0],
        [1, 1, 1, 1, 1, 1, 1]])

## Batch의 주요 구성 요소

### **1. input_ids**
- 각 텍스트 시퀀스를 토큰화하여 모델의 어휘(vocabulary) 사전에 해당하는 ID로 변환한 결과입니다.

#### 예시:
- `"Hello, world!"` → `[101, 7592, 1010, 2088, 999, 102, 0]`
  - `[101]`: 문장의 시작 토큰.
  - `[7592, 1010, 2088]`: `"Hello,"`, `"world!"` 단어의 토큰 ID.
  - `[102]`: 문장의 끝 토큰.
  - `[0]`: 패딩 토큰.
- `"Hugging Face is amazing!"` → `[101, 17662, 2227, 2003, 6429, 999, 102]`
  - 패딩 없이 길이가 같음.

---

### **2. attention_mask**
- **`1`**: 실제 데이터가 있는 부분.
- **`0`**: 패딩된 부분을 나타냅니다.

#### 예시:
- `"Hello, world!"` → `[1, 1, 1, 1, 1, 1, 0]` (마지막 `0`은 패딩을 의미).
- `"Hugging Face is amazing!"` → `[1, 1, 1, 1, 1, 1, 1]` (패딩 없음).

---

### **정리**
- **`input_ids`**: 모델이 이해할 수 있는 숫자 형태로 변환된 토큰 ID.
- **`attention_mask`**: 패딩 여부를 나타내는 마스크로, 모델이 패딩된 부분을 무시하도록 돕습니다.
